### ISSUE TITLE
dialog modify events not getting figured

### ISSUE THREAD

**Issue by jonastelzio** [davehorton/drachtio-srf#59]
Hey Dave, I hope all is well!

In https://github.com/davehorton/drachtio-srf/blob/master/lib/dialog.js L360-366 there are a bunch of tests being made which I'm trying to understand.

Drachtio doesn't fire 'modify' on the dialog when I unhold a call, because line 364 resolves to true (see if statement on line 379) - and instead it just does whatever the the default action is... which seems to be just return the currently last known SDP?

**Comment by davehorton** [comment]
Yes, I am trying to simplify handling of a few re-INVITE scenarios for the developer:

- a session refreshing re-INVITE.  Generally, there is no need for the app to do anything about this, but it may want to be notified.
- a session that has been placed on hold by the far end.  Ditto.
- a session that has been taken off hold (audio restored) by the far end.  Ditto

In those cases, the Dialog object emits 'refresh', 'hold', and 'unhold' events, respectively.

The thing in the code about redacting the o= line is that the version in the o= line will be incremementing, and we want to eliminate that from our comparisons when seeing what changed.

If none of the above are true with respect to a re-INVITE, then it is an INVITE with a new SDP that the application will want to (need to) handle, and we emit a 'modify' event in that case with `(req, res)` so the app can inspect the new sdp, do whatever and respond.

Does that help?  Is there something you need in your handling above and beyond this?

**Comment by jonastelzio** [comment]
Aah, yeah okay that makes sense.

In my case, if lets say a mobile device is being invited, so the ANSWER is being stored, then on the first HOLD, modify triggers, but is then the local SDP is set to the content of the INVITE that issues the hold. The following UNHOLD then triggers the unhold event I guess?

Is there some fancy way to disable this, and just throw everything into the modify event? Right now I'd have to subscribe to both modify, hold and unhold to essentially do the same thing

**Comment by davehorton** [comment]
I want to make sure I can understand your scenario, because you say "on the first HOLD, modify triggers.."  Do you mean that the call is established, then you get a re-INVITE on hold _but_ the 'modify' event is triggered, rather than the 'hold' event?  Because that would be a bug.

I want to make sure that the current logic is currently classifying invites-on-hold and invites-off-hold, because it almost sounds like you are experiencing an error/bug related to that

**Comment by jonastelzio** [comment]
So the flow of what happens

```
Call establishment:
Asterisk - INVITE (SDP) ->   Drachtio   - INVITE (SDP) -> Device
Asterisk <-  OK (SDP)   <-   Drachtio   <-   OK (SDP)   - Device

Then HOLD ('modify' is triggered):
Asterisk <- (Re)INVITE (SDP) <- Drachtio <- (Re)INVITE (SDP) - Device
Asterisk --  OK (SDP)   -->     Drachtio    --- OK (SDP) ----> Device

Then UNHOLD ('unhold'ø' is triggerd):
Asterisk <- (Re)INVITE (SDP) <- Drachtio <- (Re)INVITE (SDP) - Device
Asterisk --  OK (SDP)   -->     Drachtio    --- OK (SDP) ----> Device
```
From looking at the code, I think it's because the SDP returned during call establishment (200OK) is different from the SDP the device generates when it sends out a re-invite to hold the call - meaning modify gets triggered on the first hold, but because the re-invite to unhold is near identical to the hold SDP, then the local triggers the unhold event on the unhold-event (and on all subsequent re-invites from the device, presumably the hold event would also fire).

**Comment by jonastelzio** [comment]
All this being said, it's not impossible that my drachtio code is a bad. I'm battling all sorts of issues with hold/unhold right now where it straight up doesn't work when the call is initiated from the device - I think it's something with the from/to tags not being correct and RTP Engine not doing what it's supposed to...

And I've got this annoying bug where some times (like 50%) when I unhold a call established from the Asterisk side, then the call audio doesn't reestablish correctly either...

**Comment by davehorton** [comment]
can you send me (separately to daveh at drachtio.org) the drachtio log trace (at info level) from the scenario above, showing the INVITE, the 'hold', and then the 'unhold'?

**Comment by davehorton** [comment]
There are sort of 2 problems here:

1. We weren't smart enough to detect the first re-INVITE was an invite on hold, so we emitted 'modify'
2. We then *did* detect that the next re-invite was an invite off hold, but emitting an 'unhold' when we never emitted 'hold' previously led to problems.

First thing, I have fixed #2.  So this means you will now get 2 'modify' events rather than 'modify' then 'unhold'.

To test, can you `npm install drachtio-srf@beta.` ?

**Comment by jonastelzio** [comment]
Alright I've taken the beta version for a spin. It works as expected for the first hold/unhold, but for subsequent hold/unholds it starts firing the hold/unhold events, instead of modify. I don't know if this is intended?

This is what happens:

1. HOLD: modify fired
2. UNHOLD: modify fired
3: HOLD: hold fired
4: UNHOLD: unhold fired

I'm in doubt on how I'm supposed to implement the hold/unhold events, if I was to use them. In dialog.js, we've got the following code:

`        if (refresh) {
          this.emit('refresh', req);
        }
        else if (hold) {
          this.local.sdp = this.local.sdp.replace(/a=sendrecv\r\n/g, 'a=recvonly\r\n') ;
          this.emit('hold', req) ;
        }
        else if (unhold) {
          this.local.sdp = this.local.sdp.replace(/a=recvonly\r\n/g, 'a=sendrecv\r\n') ;
          this.emit('unhold', req) ;
        }
        if ((refresh || hold || unhold) || (modify && 0 === this.listeners('modify').length)) {
          debug('responding with 200 OK to reINVITE') ;
          res.send(200, {
            body: this.local.sdp,
            headers: {
              'Contact': this.local.contact,
              'Content-Type': 'application/sdp'
            }
          }) ;
        }
        else if (modify) {
          this.emit('modify', req, res) ;
        }
`

When it fires hold/unhold it does some modification on the localsdp based on a stringe replacement, and then that result is returned to the UAC? I need to pass this SDP through RTP Engine, pass it to Asterisk, and then pass Asterisk's answer through RTP Engine, and then return that to the UAC.

**Comment by davehorton** [comment]
Right, so you need a feature to completely disable the hold/unhold features.  I will wok on that next

**Comment by jonastelzio** [comment]
What about if it only do this internal hold/unhold logic if those events are actually subscribed?

Downside I guess is that it would change the current behaviour for those who rely on this logic even if they're not subscribed for those events. They don't appear to be in the documentation though?

**Comment by davehorton** [comment]
Yes, I kind of like that approach

**Comment by davehorton** [comment]
addressed in 4.4.19
